packages/venia-ui/lib/components/StoreCodeRoute/storeCodeRoute.js

```diff
@@ -0,0 +1,55 @@
+import { useEffect } from 'react';
+import { useHistory, useRouteMatch } from 'react-router-dom';
+import { BrowserPersistence } from '@magento/peregrine/lib/util';
+
+const storage = new BrowserPersistence();
+
+/**
+ * This component checks for use of a store code in the url that is not the
+ * current base. If found, it updates the local storage values for code/currency
+ * and reloads the page so that they are used in the graphQL headers.
+ */
+const StoreCodeRoute = () => {
+    const history = useHistory();
+
+    const storeCodes = [];
+    const storeCurrencies = {};
+
+    AVAILABLE_STORE_VIEWS.forEach(store => {
+        storeCodes.push(store.code);
+        storeCurrencies[store.code] = store.default_display_currency_code;
+    });
+
+    const pattern = `/:storeCode(${storeCodes.join('|')})?`;
+    const match = useRouteMatch(pattern);
+
+    useEffect(() => {
+        // The current store code won't be matched as it's included as the
+        // basename, if we match another store we need to change the current
+        // store.
+        if (match.params.storeCode) {
+            // Only execute if one store code is present in the URL, multiple
+            // store codes will cause the store state to break and cause weird
+            // side effects for the user.
+            const regex = new RegExp(`\/(${storeCodes.join('|')})`, 'g');
+            const storeCodesInUrl = window.location.pathname.match(regex);
```

Comment:
I am not sure if this is working as expected.

Try on this JS snippet

```es6
const storeCodes = ['en', 'fr', 'en-uk', 'frs'];
const pattern = `/:storeCode(${storeCodes.join('|')})?`;
const regex = new RegExp(`\/(${storeCodes.join('|')})`, 'g');

'/frs/home.html'.match(regex) // should be ['/frs'] but returns ['/fr']
'/en-uk/home.html'.match(regex) // should be ['en-uk'] but returns ['/en']
```

Let me know if that is expected behavior.

Reply:
Good catch!

Reply:
@davemacaulay suggested ordering `storeCodes` by length to solve this issue for now.